### ISSUE TITLE
unused glyph <Delete> imported into nested substitution subtables

### ISSUE THREAD

**Issue by MaddTheSane** [fontforge/fontforge#4190]
I found some old Copland install stuff. Included was a font suitcase called Tekton Plus. Tekton Plus would have been used to show off the QuickDraw GX font stuff, including glyphs for small caps, super- and subscript. While the font itself does work under Mac OS X, trying to load it into FontForge ("Nifty, you've got one of the old Apple/Adobe type1 sfnts here") then export it out results in it complaining:

> Lookup subtable contains unused glyph <Delete> making the whole subtable invalid
>Lookup subtable contains unused glyph <Delete> making the whole subtable invalid
>Lookup subtable contains unused glyph <Delete> making the whole subtable invalid

This does make me wonder: Could the font be using undocumented Apple typography methods, or does FontForge not know how to handle certain documented AAT calls.

Built using the latest master, specifically af998cc28fc69ccce56ffbaf2c62aa57d2790bef.

**Comment by skef** [comment]
Is the expectation behind this issue that files made 25 years ago will export into modern formats without any editing or modification? Or I guess more specifically that if the formats are fully compatible in some broad sense they should do so?

We can keep this kind of issue open but I'm afraid the only result will be to continue to have hundreds and hundreds of open issues. 

However, unless you're *willing to attach the file* I am just going to close. How on earth would we ever pursue a fix without it? Sending it privately doesn't work when it will likely be years before anyone is able to look at this.

**Comment by MaddTheSane** [comment]
Attaching file.
[TektonPlus.dfont.zip](https://github.com/fontforge/fontforge/files/4248781/TektonPlus.dfont.zip)
The file has been converted to a dfont. Other than that, it is identical to the one I found.

**Comment by skef** [comment]
"\<Delete\>" is being added via the `MAC_DELETED_GLYPH_NAME` define in `ttf.h`. This is in virtue of `TTF_SetMortSubs()` being called with `gsubs == 0xffff`, which is treated as a specialized glyph number. The code that reads the value in `read_perglyph_subs()` is:
```
        here = ftell(ttf);
        subs = getushort(ttf);
        if ( subs>=info->glyph_cnt && subs!=0xffff )    /* 0xffff means delete the substituted glyph */
```
Which means we're actually reading the value and it is anticipated.

`tottfaat.c:aat_dumpmorx_substitutions()` has the lines
```
                    } else if ( msc==NULL &&
                            strcmp(pst->u.subs.variant,MAC_DELETED_GLYPH_NAME)==0 ) {
                        maps[gcnt++] = 0xffff;
```
while `lookups.c:ApplySingleSubsAtPos()` has the lines:
```
if ( strcmp(pst->u.subs.variant,MAC_DELETED_GLYPH_NAME)==0 ) {
        /* Under AAT we delete the glyph. But OpenType doesn't have that concept */
        int i;
        for ( i=pos+1; i<data->cnt; ++i )
            data->str[pos-1] = data->str[pos];
        --data->cnt;
return( pos );
```
However, this is the only type of substitution with a work-around. The tables that aren't working aren't single substitutions. For example, the sfd entry `<14,0> Hyphens to Em Dash lookup 17 nested-substitutions 1 subtable" <Delete>` is a subtable in a contextual substitution. 

In short, this issue is a request to extend workarounds for the ATT "deleted glyph" semantic to other OpenType substitution types.
